Add GT.M version of Docker ViViaN

josephsnyder opened the pull request:
Add the local VistA option for a GT.M/YottaDB install and pass those
arguments on to the ViViaN install.  Remove all instance information from
the CMakeCache information.

Instructions for the GT.M have been added to a Gerrit change here:  http://review.code.osehra.org/#/c/1045/

josephsnyder committed:
Add GT.M version of Docker ViViaN

Add the local VistA option for a GT.M/YottaDB install and pass those
arguments on to the ViViaN install.  Remove all instance information from
the CMakeCache information.

ChristopherEdwards commented on ViViaN/vivianInstall.sh:
```diff
@@ -9,11 +9,12 @@ usage()
       -h    Show this message
       -s    Script Directory
       -i    Instance name (Namespace/Database for Caché)
+      -y    Building a non-Cache instance
```
This seems weird. I'd say a GT.M or YottaDB instance?

ChristopherEdwards commented on ViViaN/vivianInstall.sh:
```diff
@@ -26,6 +27,9 @@ do
             instance=$(echo $OPTARG |tr '[:upper:]' '[:lower:]')
             ;;
+        y)
+            nonCache=$OPTARG
```
yottadb?

ChristopherEdwards commented on ViViaN/vivianInstall.sh:
```diff
@@ -41,40 +45,67 @@ if [[ -z $scriptdir ]]; then
     scriptdir=/opt/vista
 fi
 
+if [[ -z $nonCache ]]; then
```
ditto

ChristopherEdwards commented on ViViaN/vivianInstall.sh:
```diff
@@ -41,40 +45,67 @@ if [[ -z $scriptdir ]]; then
 # cp $scriptdir/ViViaN/viv.conf /etc/httpd/conf.d
-basedir=/opt/cachesys/$instance
+
+if ! $nonCache; then
```
the double negative here is confusing. I'd flip this around since there is an else statement attached.

ChristopherEdwards commented on ViViaN/vivianInstall.sh:
```diff
@@ -41,40 +45,67 @@ if [[ -z $scriptdir ]]; then
 curl -fsSL --progress-bar https://foia-vista.osehra.org/VistA_Integration_Agreement/2018_January_22_IA_Listing_Descriptions.TXT -o ICRDescription.txt
 echo "Downloading OSEHRA VistA Testing Repository"
-curl -fsSL --progress-bar https://github.com/OSEHRA/VistA/archive/master.zip -o VistA-master.zip
+curl -fsSL --progress-bar https://github.com/josephsnyder/VistA/archive/fix_menus_dir_creation.zip -o VistA-master.zip
```
This should probably point to a OSEHRA repo instead of a personal fork

josephsnyder replied:
Sorry, this wasn't supposed to be a part of the commit.  I'll fix this

ChristopherEdwards commented on ViViaN/vivianInstall.sh:
```diff
@@ -41,40 +45,67 @@ if [[ -z $scriptdir ]]; then
 unzip -q VistA-master.zip
 rm VistA-master.zip
-mv VistA-master VistA
+mv VistA-fix_menus_dir_creation VistA
```
This shouldn't occur if moved to an osehra repo and the change is merged to master.

ChristopherEdwards commented on ViViaN/vivianInstall.sh:
```diff
@@ -41,40 +45,67 @@ if [[ -z $scriptdir ]]; then
 mkdir -p /opt/VistA-M/Packages
 cp /opt/VistA/Packages.csv /opt/VistA-M/
 
+if $nonCache; then
```
ditto

ChristopherEdwards commented on autoInstaller.sh:
```diff
@@ -297,12 +301,14 @@ if $installRPMS; then
 fi
 
 if $installgtm || $installYottaDB ; then
+    gtmInst=true
```
this seems odd, just use the existing environment variables $installgtm || $installYottaDB

ChristopherEdwards commented on autoInstaller.sh:
```diff
@@ -297,12 +301,14 @@ if $installRPMS; then
 fi
 
 if $installcache; then
+    gtmInst=false
```
you can check for the absence of $installgtm/$installYottaDB instead of creating this variable

ChristopherEdwards commented on autoInstaller.sh:
```diff
@@ -335,84 +341,104 @@ if $installgtm || $installYottaDB; then
-    done
-    echo "Done compiling routines"
+if ($installgtm || $installYottaDB); then
+  # Build a dashboard and run the tests to verify installation
```
This comment isn't valid here, should be further down where we do that part

ChristopherEdwards commented on autoInstaller.sh:
```diff
@@ -335,84 +341,104 @@ if $installgtm || $installYottaDB; then
+      echo "Using local files found in ./GTM/"
+      cd $scriptdir
+      #
+      unzip -q ./GTM/VistA.zip -d /tmp/gtmout
```
Where did this file come from? This logic just seems weird in general:

somewhere got a new zipfile, we extract it, move the db and routines into place, then modify the existing GDE to point to the new database?

josephsnyder replied:
The new VistA.zip is supposed to be supplied by the user, which is described in the OSEHRA documentation for this part of the script.  

That is the basis of it.  From what I think i understood about GDE, you cannot have two regions with "*" as the global name so I had to drop the default file to make our incoming one read by GT.M.  Should I have tried to import them somehow?

ChristopherEdwards replied:
I'm going to seek some clarification on this as it still seems out of place to me.

1. Adding the zip file of VistA seems to be trying to mimic what the Caché scripts do. The Caché scripts are rather special in that it assumes that you'll get the CACHE.DAT to start. That isn't what I really want the Caché scripts to do in the end, but that is what others who contacted me to do the work wanted. This addition seems to be a new feature rather than using the existing framework (which should require a different flag instead of generateViDox). Why can't you use the pending installation of Routines/Globals to do the work?

2. Can you add a little more documentation to this readme about how this works with GTM/YottaDB. I don't want users to have to keep jumping documentation to figure out how to do things. To me this is a separation of concerns issue: the documentation about how to use the vivian/dox flags to autoInstaller.sh should be documented here. Documentation about vivian/dox in gerneral should be on the osehra pages for vivian/dox.

You are correct that two regions can't have "*" as the global name. I got confused as to what is going on here as this is the section about installing VistA on YottaDB/GTM and wondered why a Vivian/Dox flag is modifying it.

ChristopherEdwards replied:
If you are adding the feature that one can use a "binary" of the GT.M/YottaDB database it needs to be made clear that the gld needs to come with it.

josephsnyder replied:
Thanks Christopher,

I'm sorry about the huge delay but I think that I'm ready to push this forward and get this merged in.  
Regarding #1: the point of the script is to allow someone to add their own instance into the ViViaN environment.  I figured the best way to do that consistently would be to expect a "binary" of the MUMPS environment and then put that in when the instance was built in the original script.  Since the Cache instance only has the one install path, it was easy enough to just utilize it again.  I figured GT.M/Yotta would be the same to only install it once but I understand the reasoning behind the separation.  I have since moved the "install" of the local files from the autoInstaller into the vivianInstall script and have now expected the .gld file to come along with the installer.

Regarding #2:  I have added the information from the ViViaN documentation into the README.  

Please take a look at the force-pushed update and let me know if there is anything else that you feel is not up to standards

ChristopherEdwards replied:
Is looking at this next week ok?

josephsnyder replied:
Fine by me.

ChristopherEdwards commented on autoInstaller.sh:
```diff
@@ -452,7 +478,7 @@ fi
   if $installgtm || $installYottaDB; then
     su $instance -c "source $basedir/etc/env && pushd $scriptdir && $postInstallScript && popd"
-  elif $installcache; then
+  elif $installcache; then  
```
trailing whitespace

ChristopherEdwards commented on autoInstaller.sh:
```diff
@@ -467,5 +493,5 @@ fi
 # Generate ViViaN Documentation
 if $generateViVDox; then
-    $scriptdir/ViViaN/vivianInstall.sh -i $instance -s $scriptdir
+    $scriptdir/ViViaN/vivianInstall.sh -i $instance -s $scriptdir -y $gtmInst
```
i think this could be resolved for looking for the gtm/yottadb environment variables by default and leaving the -y as an optional parameter for those installs.

josephsnyder replied:
I'm going to remove the gtmInst variable and instead look for the presence of the env file for GT.M/YottaDB as the trigger for GT.M instances

if [ -f /home/$instance/etc/env]; then
  \<do gtm things\>
else
  \<do cache things\>

josephsnyder committed:
Add fixes from review

Remove unnecessary gtmInst argument to the vivianInstall file

Update comments and variable names

josephsnyder commented:
@ChristopherEdwards,

I think I've pushed the changes that you've requested.  PTAL when you get the chance.
